modules/qna/src/views/list/ElementsList.jsx

```diff
@@ -0,0 +1,82 @@
+import React from 'react'
+import { ListGroupItem, Glyphicon } from 'react-bootstrap'
+import { InputElement } from './InputElement'
+import style from './style.scss'
+
+export class ElementsList extends React.Component {
+  elementInputRef = React.createRef()
+  state = {
+    inputValue: '',
+    error: undefined,
+    elements: [],
+    editElementIndex: undefined
+  }
+
+  createElement = element => {
+    this.setState({ elements: [...this.state.elements, element] }, this.onElementsChange)
+  }
+
+  updateElement = (element, index) => {
+    const elements = this.state.elements
+    if (elements[index]) {
+      elements[index] = element
+      this.setState({ elements, editElementIndex: undefined }, this.onElementsChange)
+    }
+  }
+
+  deleteElement = index => {
+    const elements = this.state.elements
+
+    if (elements[index]) {
+      elements.splice(index, 1)
+      this.setState({ elements }, this.onElementsChange)
```

Comment:
not that it's an issue at all but just so you know, it's not necessary to wait the state change to trigger the `onElementsChange` callback. You already computed the new state of elements why not straight call it ?

```
this.setState({elements})
this.props.onElementsChange && this.props.onElementsChange(elements)
```
if you really, badly want to use your `onElementsChange` method, simply add elements as parameter
it's just a suggestion it doesn't really change anything

Reply:
`onElementsChange` will return state.elements to its parent and `setState` is async, so if I want to be sure the elements are up-to-date I have to do it this way. I'm wrapping `props.onElementsChange()` in a method so that I can easily reference it. Also in case the prop name changes, you just have to rename it in one place.